packages/components/src/custom-select-control/index.tsx

```diff
@@ -51,7 +51,9 @@ function getDescribedBy( currentValue: string, describedBy?: string ) {
 	return sprintf( __( 'Currently selected: %s' ), currentValue );
 }
 
-function CustomSelectControl( props: CustomSelectProps ) {
+function CustomSelectControl< T extends CustomSelectOption >(
```

Comment:
The generic param is not going to be set manually by user. It's used for type inferring. To be specific, it will use the item type of the `options` prop automatically. So consumers would not need to change anything.

You could checkout this branch and test with the snippets that pasted in the "Testing Instructions" section.

Reply:
TypeScript should always be able to infer the `T` type. `CustomSelectProps` have a mandatory `options: Array<T>` field, so `T` is always the type of whatever you pass as `options`, it's easy to infer. TypeScript will merely check if that type is compatible with `CustomSelectOption`.